File: Codes/process.php
<?php
// Path to the GNSS files
$folderPath = "../files_stations/";

// Get all .txt files from the folder
$files = glob($folderPath . "*.txt");

// Data storage arrays
$results = [];
$stationInfo = []; // Store station information
$periods = []; 
$proc = "";    // Store processing 

// Iterate over each file
foreach ($files as $file) {
    $fileName = basename($file);  // Get the base name of the file
    $fileContent = file($file, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES); // Read file content, ignore empty lines
    if (!is_dir($folderPath)) {
        die("The specified folder does not exist.");
    }
    
    if (empty($files)) {
        die("No files found in the folder.");
    }

    // Variables to store header information
    $nodeCode = $nodeName = $nodeUrl = null;
    $nodeLat = $nodeLon = $nodeElevation = null;

    // Read the file's header section
    foreach ($fileContent as $line) {
        if (strpos($line, "PROC:") !== false) {
            // Extract the processing information (PROC)
            $proc = trim(explode(":", $line, 2)[1]);
            $proc = str_replace(['{', '}'], '', $proc);  // Clean up the PROC value
        } elseif (strpos($line, "NODE_FID:") !== false) {
            $nodeCode = trim(explode(":", $line, 2)[1]);  // Extract the node code
        } elseif (strpos($line, "NODE_NAME:") !== false) {
            $nodeName = trim(explode(":", $line, 2)[1], ' "');  // Extract node name
        } elseif (strpos($line, "NODE_LATITUDE:") !== false) {
            $nodeLat = floatval(trim(explode(":", $line, 2)[1]));  // Extract latitude
        } elseif (strpos($line, "NODE_LONGITUDE:") !== false) {
            $nodeLon = floatval(trim(explode(":", $line, 2)[1]));  // Extract longitude
        } elseif (strpos($line, "NODE_ELEVATION:") !== false) {
            $nodeElevation = floatval(trim(explode(":", $line, 2)[1]));  // Extract elevation
        } elseif (strpos($line, "NODE_URL:") !== false) {
            $nodeUrl = trim(explode(":", $line, 2)[1]);  // Extract node URL
        } elseif (strpos($line, "TIME_PERIODS:") !== false) {
            // Extract time periods and store them as an array of integers
            $periods = array_map('intval', explode(',', trim(explode(":", $line, 2)[1])));
        } elseif (strpos($line, "yyyy mm dd") !== false) {
            // Column names for the GNSS data
            $columnNames = preg_split('/\s+/', trim($line));  // Split the date column names
            break; // End of header processing
        }
    }

    // Store station information
    $stationInfo[$fileName] = [
        "code" => $nodeCode,
        "name" => $nodeName,
        "latitude" => $nodeLat,
        "longitude" => $nodeLon,
        "elevation" => $nodeElevation,
        "url" => $nodeUrl
    ];

    // Process GNSS data lines
    foreach ($fileContent as $line) {
        // Skip comment lines
        if (strpos($line, "#") === 0) continue;

        // Split the line into individual values
        $values = preg_split('/\s+/', trim($line));
        if (count($values) < count($columnNames)) continue;  // Ensure the line has all expected values

        // Extract year, month, day, hour, minute, second for the date
        list($year, $month, $day, $hour, $minute, $second) = array_slice($values, 0, 6);
        // Format the date as YYYY-MM-DD
        $date = sprintf("%04d-%02d-%02d", $year, $month, $day);

        // Initialize the date in the results array if not already present
        if (!isset($results[$date])) {
            $results[$date] = [];
        }

        // Store vectors and errors for each period
        $vectors = [];
        foreach ($periods as $index => $days) {
            $indexOffset = 6 + $index * 6;  // Calculate the offset for each period's data
            $vectors[$days] = [
                "vector" => array_map('floatval', array_slice($values, $indexOffset, 3)), // Extract vector values
                "error"  => array_map('floatval', array_slice($values, $indexOffset + 3, 3))  // Extract error values
            ];
        }

        // Store the vectors and the station position for the current date
        $results[$date][$fileName] = [
            "vectors" => $vectors,
            "position" => [
                "lat" => $nodeLat,
                "lon" => $nodeLon,
                "elevation" => $nodeElevation
            ]
        ];
    }
}

// Return the JSON response containing station info, periods, and data
echo json_encode([
    "proc" => $proc,
    "stations" => $stationInfo,
    "periods" => $periods,
    "data" => $results
]);
?>

File: Codes/process.pl
#!/usr/bin/perl
use strict;
use warnings;
use JSON;
use File::Basename;
use CGI;
print "Content-Type: application/json\n\n";

# Chemin vers les fichiers GNSS
my $folder_path = "../Fichiers_stations/";

# Liste des fichiers dans le dossier
my @files = glob("$folder_path/*.txt");
die encode_json({ "error" => "Aucun fichier trouvé dans le dossier." }) if scalar @files == 0;

# Variables de stockage
my %results;
my %station_info;
my @periods;
my $proc = "";
my @column_names;

foreach my $file (@files) {
    my $file_name = basename($file);
    
    open my $file_handle, '<', $file or die encode_json({ "error" => "Impossible d'ouvrir le fichier: $file_name" });
    my @file_content = grep { $_ ne "" } <$file_handle>;
    close $file_handle;

    # Variables pour les métadonnées
    my ($node_code, $node_name, $node_url, $node_lat, $node_lon, $node_elevation) = (undef, undef, undef, undef, undef, undef);

    # Lecture de l'en-tête
    foreach my $line (@file_content) {
        $line = trim($line);
        if ($line =~ /PROC:/) {
            $proc = extract_value($line);
            $proc =~ s/[{}]//g;
        } elsif ($line =~ /NODE_FID:/) {
            $node_code = extract_value($line);
        } elsif ($line =~ /NODE_NAME:/) {
            $node_name = extract_value($line);
        } elsif ($line =~ /NODE_LATITUDE:/) {
            $node_lat = $line =~ /NODE_LATITUDE:/ ? 0 + extract_value($line) : undef;
        } elsif ($line =~ /NODE_LONGITUDE:/) {
            $node_lon = $line =~ /NODE_LONGITUDE:/ ? 0 + extract_value($line) : undef;
        } elsif ($line =~ /NODE_ELEVATION:/) {
            $node_elevation = $line =~ /NODE_ELEVATION:/ ? 0 + extract_value($line) : undef;
        } elsif ($line =~ /NODE_URL:/) {
            $node_url = extract_value($line);
        } elsif ($line =~ /TIME_PERIODS:/) {
            @periods = map { int($_) } split(',', extract_value($line));
        } elsif ($line =~ /yyyy mm dd/) {
            @column_names = split(/\s+/, $line);
            last;
        }
    }

    # Stockage des infos de la station
    $station_info{$file_name} = {
        "code"      => $node_code,
        "name"      => $node_name,
        "latitude"  => $node_lat,
        "longitude" => $node_lon,
        "elevation" => $node_elevation,
        "url"       => $node_url
    };

    # Lecture des données GNSS
    foreach my $line (@file_content) {
        $line = trim($line);
        next if $line =~ /^#/; # Ignorer les commentaires
        next if $line =~ /yyyy mm dd/; # Ignorer l'en-tête

        my @values = split(/\s+/, $line);
        next if scalar @values < scalar @column_names; # Vérification du format

        my ($year, $month, $day, $hour, $minute, $second) = @values[0..5];
        my $date = sprintf("%04d-%02d-%02d", $year, $month, $day);

        $results{$date} //= {}; # Création si inexistant

        my %vectors;
        foreach my $index (0..$#periods) {
            my $index_offset = 6 + $index * 6;
            my @vector = @values[$index_offset..$index_offset + 2];
            my @error = @values[$index_offset + 3..$index_offset + 5];
            
            # Remplacer "NaN" par 0 et convertir en nombre
            for my $i (0..2) {
                $vector[$i] = ($vector[$i] eq "NaN") ? 0 : 0 + $vector[$i];  # Convertir en nombre
                $error[$i]  = ($error[$i] eq "NaN")  ? 0 : 0 + $error[$i];   # Convertir en nombre
            }

            $vectors{$periods[$index]} = {
                "vector" => \@vector,
                "error"  => \@error
            };
        }

        $results{$date}{$file_name} = {
            "vectors" => \%vectors,
            "position" => {
                "lat"       => $node_lat,
                "lon"       => $node_lon,
                "elevation" => $node_elevation
            }
        };
    }
}

# Sortie JSON
print encode_json({
    "proc"      => $proc,
    "stations"  => \%station_info,
    "periods"   => \@periods,
    "data"      => \%results
});

# Fonctions utilitaires
sub trim {
    my ($string) = @_;
    $string =~ s/^\s+|\s+$//g;
    return $string;
}

sub extract_value {
    my ($line) = @_;
    return trim((split(":", $line, 2))[1]);
}
